### ISSUE TITLE
Time precision problem

### ISSUE THREAD

**Issue by dr-itz** [jruby/jruby#5558]
### Environment

* JRuby 9.1.17.0, 9.2.5.0
* JDK:
    ```
    java version "1.8.0_192"
    Java(TM) SE Runtime Environment (build 1.8.0_192-b12)
    Java HotSpot(TM) 64-Bit Server VM (build 25.192-b12, mixed mode)
    ```

### Problem

Running AR-JDBC tests, we have intermittent failures in Rails Time precision tests. See https://github.com/jruby/activerecord-jdbc-adapter/pull/979 for the test exclusion

Extracted test follows. MRI (tested with 2.4.4 and 2.5.1) runs w/o failures, JRuby fails usually between 1000-2000 iterations. 

### Reproducer

The script contains all relevant parts from Rails (5.2.2 but 5.0.x is affected too), no dependencies

```ruby
# what active_record's Time.change does internally
def change_time_usec(time, new_usec)
  new_sec = time.sec
  new_sec += Rational(new_usec, 1000000)
  Time.new(time.year, time.month, time.day, time.hour, time.min, new_sec)
end

def change_time_nsec(time, new_nsec)
  change_time_usec(time, Rational(new_nsec, 1000))
end

# from 'activemodel/lib/active_model/type/helpers/time_value.rb'
def apply_seconds_precision(value, precision)
  number_of_insignificant_digits = 6 - precision
  round_power = 10**number_of_insignificant_digits
  new_usec = value.usec - value.usec % round_power
  change_time_usec(value, new_usec)
end

# basically what
#   activerecord/test/cases/date_time_precision_test.rb
# does in #test_datetime_precision_is_truncated_on_assignment
# note: need to repeat and sleep to reproduce, but not 100% reliable even then
5000.times do |i|
  now = Time.now
  time = change_time_nsec(now, 123456789)
  time2 = apply_seconds_precision(time, 6)

  unless time2.nsec == 123456000
    puts "fail after #{i} iterations"
    puts now.strftime "%H:%M:%S.%N"
    puts time.strftime "%H:%M:%S.%N"
    puts time2.strftime "%H:%M:%S.%N"
    break
  end

  sleep 0.01
end
```

**Comment by dr-itz** [comment]
Ok, 100% reliable reproducer:

```ruby
def time_change_sec_usec(time, sec, new_usec)
  # problem is the addition here!
  new_sec = sec + Rational(new_usec, 1000000)
  Time.new(time.year, time.month, time.day, time.hour, time.min, new_sec)
end

now = Time.now
for i in 0..59 do
  time = time_change_sec_usec(now, i, 123456)
  unless time.nsec == 123456000
    puts "fail #{i}: " + time2.strftime("%H:%M:%S.%N")
  end
end
```

this prints:

```
fail 8: 23:43:08.123455999
fail 32: 23:43:32.123455999
fail 33: 23:43:33.123455999
fail 34: 23:43:34.123455999
```

@kares this actually starts to make sense in terms of a rounding problem, no?

**Comment by kares** [comment]
indeed it does, will need to look down the path ... would no rely on `strftime("%H:%M:%S.%N")` for [now](https://github.com/jruby/jruby/issues/5556)
btw. at some point I added a similar piece of AS to test time-change: https://github.com/jruby/jruby/blob/master/test/jruby/test_time.rb#L96

**Comment by dr-itz** [comment]
@kares I did a little digging myself. Was an easy fix 😃
